### PR TITLE
feat: adding json, yaml and toml sources

### PR DESCRIPTION
**Pull request by Smixi**
feat: https://github.com/pydantic/pydantic-settings/issues/210

Selected Reviewer: @samuelcolvin

### CODE REVIEW

**Comment by samuelcolvin** [comment]
I think you mean toml, not JSON - please update the title.

**Comment by hramezani** [comment]
Thanks @Smixi for the PR 🙏 

A couple of things to address:

- If the dependency is installed, we need to add the source class by default in `settings_customise_sources`. So, no need to override the `settings_customise_sources` to use the new source classes. They should have less priority than the existing sources to be backward-compatible.
- CI is broken for Python 3.12. it seems the `pyyaml 6.0` can't be installed. Have you tried the `pyyaml 6.0.1`?
- You've added the `JsonConfigSettingsSource` as well. not sure we want that. @samuelcolvin what do you think?

File: pydantic_settings/sources.py [hunk 1]
```diff
@@ -19,10 +20,48 @@
 from pydantic_settings.utils import path_type_label
 
 if TYPE_CHECKING:
+    if sys.version_info.minor >= 11:
```
Comment (hramezani) [inline pydantic_settings/sources.py:23]:
```suggestion
    if sys.version_info >= (3, 11):
```

**Comment by Smixi** [comment]
> Thanks @Smixi for the PR 🙏
> 
> A couple of things to address:
> 
> * If the dependency is installed, we need to add the source class by default in `settings_customise_sources`. So, no need to override the `settings_customise_sources` to use the new source classes. They should have less priority than the existing sources to be backward-compatible.
> * CI is broken for Python 3.12. it seems the `pyyaml 6.0` can't be installed. Have you tried the `pyyaml 6.0.1`?
> * Documentation has to be added in https://github.com/pydantic/pydantic-settings/blob/main/docs/index.md
> * You've added the `JsonConfigSettingsSource` as well. not sure we want that. @samuelcolvin what do you think?

If they are default loaded, it means the default settings class is in settings_customise_sources, which will result in the API breaking (people having custom class order will fail) or does it make sense ?

**Comment by hramezani** [comment]
Here is the source settings priority right now:

- init_settings
- env_settings
- dotenv_settings
- file_secret_settings

So, if we add new sources at the end there shouldn't be a breaking change and there will be new sources that users can use by defining the file path and encoding in config.

It means that releasing the new `pydantic-settings` won't break existing settings classes and users can benefit from new sources by defining the required configs.

**Comment by Smixi** [comment]
My concerns is more about user having their own custom orders. They will need to reimplement the `settings_customise_sources` to add the new parameters (like I did in the tests) ?
Pyaml 6.0.1 seems to fix the issue

**Comment by hramezani** [comment]
> My concerns is more about user having their own custom orders. They will need to reimplement the settings_customise_sources to add the new parameters (like I did in the tests) ?

That's a good point. didn't think about this.
So, please revert the change and document how to use the new sources. sorry for this

**Comment by Smixi** [comment]
@hramezani  That should be good. I just don't know how to deal with the coverage.

File: pydantic_settings/sources.py [hunk 2]
```diff
@@ -664,6 +704,103 @@ def __repr__(self) -> str:
         )
 
 
+class ConfigFileSourceMixin(ABC):
+    def _read_files(self, files: PathType | None) -> dict[str, Any | None]:
+        if files is None:
+            return {}
+        if isinstance(files, (str, os.PathLike)):
+            files = [files]
+        vars: dict[str, Any | None] = {}
```
Comment (hramezani) [inline pydantic_settings/sources.py:713]:
```suggestion
        vars: dict[str, Any] = {}
```

File: pydantic_settings/sources.py [hunk 3]
```diff
@@ -664,6 +704,103 @@ def __repr__(self) -> str:
...
+        for file in files:
+            file_path = Path(file).expanduser()
+            if file_path.is_file():
+                vars.update(self._read_file(file_path))
+        return vars
+
+    @abstractmethod
+    def _read_file(self, path: Path) -> dict[str, Any | None]:
```
Comment (hramezani) [inline pydantic_settings/sources.py:721]:
```suggestion
    def _read_file(self, path: Path) -> dict[str, Any]:
```

File: pydantic_settings/sources.py [hunk 4]
```diff
@@ -664,6 +704,103 @@ def __repr__(self) -> str:
         )
 
 
+class ConfigFileSourceMixin(ABC):
+    def _read_files(self, files: PathType | None) -> dict[str, Any | None]:
```
Comment (hramezani) [inline pydantic_settings/sources.py:708]:
```suggestion
    def _read_files(self, files: PathType | None) -> dict[str, Any]:
```

File: pydantic_settings/sources.py [hunk 5]
```diff
@@ -664,6 +704,103 @@ def __repr__(self) -> str:
...
+            yaml_file_encoding
+            if yaml_file_encoding is not None
+            else settings_cls.model_config.get('yaml_file_encoding')
+        )
+        self.yaml_data = self._read_files(self.yaml_file_path)
+        super().__init__(settings_cls, self.yaml_data)
+
+    def _read_file(self, file_path: Path) -> dict[str, Any | None]:
+        with open(file_path, encoding=self.yaml_file_encoding) as yaml_file:
+            import_yaml()
```
Comment (hramezani) [inline pydantic_settings/sources.py:800]: Isn't it better to move `import_yaml` out of the block to check the import before opening the file?

File: pydantic_settings/sources.py (hunk 4 above)
Comment (Smixi) [reply pydantic_settings/sources.py:708]: As in env settings, I think this is allowed in config files to have null values, as it's defined in json and in yaml specs ?

**Comment by hramezani** [comment]
> @hramezani That should be good. I just don't know how to deal with the coverage.

I think it can be tested like [the old dotenv test in Pydantic V1](https://github.com/pydantic/pydantic/blob/3ddb509786089f9df0658303545ae4b66db2d47c/tests/test_settings.py#L909)

and then you can add a new step to the CI to [uninstall deps and run the tests](https://github.com/pydantic/pydantic/blob/3ddb509786089f9df0658303545ae4b66db2d47c/.github/workflows/ci.yml#L242) to run the above tests

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -56,6 +56,15 @@ jobs:
         COVERAGE_FILE: .coverage.${{ runner.os }}-py${{ matrix.python }}
         CONTEXT: ${{ runner.os }}-py${{ matrix.python }}
 
+    - name: uninstall deps
+      run: pip uninstall -y tomlkit PyYAML python-dotenv
```
Comment (hramezani) [inline .github/workflows/ci.yml:60]: `python-dotenv` is not an optional dep. no need to test for it

Comment (Smixi) [reply .github/workflows/ci.yml:60]: Should I remove it from the test then ? It was previously set here:
```python
@pytest.mark.skipif(dotenv, reason='python-dotenv is installed')
def test_dotenv_not_installed(tmp_path):
    p = tmp_path / '.env'
    p.write_text('a=b')
```

Comment (hramezani) [reply .github/workflows/ci.yml]: Yes, `python-dotenv` is no an optional dep and will be installed with `pydantic-settings`

**Review by hramezani — APPROVED** [review]

**Comment by hramezani** [comment]
LGTM. Thanks @Smixi.

@samuelcolvin I will keep this PR open. please take a look if you want.

**Comment by korolkevich** [comment]
Hi @Smixi,
Thank you for your work.

I have a question:
If we have nested YAML configuration:
`app:
  pass: pass
  login: login
`
How do we replace the default configuration with environment variables, as it is done in Dynaconf (app__pass='new_val' or app.pass='new_val'), and handle deeper configurations? I couldn't find this in your PR.

**Comment by Smixi** [comment]
@korolkevich This PR is only a loader for variables. 
I think your use case is more about using [this mecanism](https://docs.pydantic.dev/latest/concepts/pydantic_settings/#changing-priority) and declare env_settings before the yaml setting source

File: docs/index.md [hunk 7]
```diff
@@ -534,6 +534,66 @@ Last, run your application inside a Docker container and supply your newly creat
...
+        return (TomlConfigSettingsSource(settings_cls),)
+```
+
+This will be able to read the following "config.toml" file, located in your working directory:
+
+```toml
+foobar = "Hello"
+[nested]
+nested_field = "world!"
+"""
```
Comment (samuelcolvin) [inline docs/index.md:594]: I don't understand what this is?

**Review by samuelcolvin — APPROVED** [review]
otherwise LGTM.

Thank you so much.

**Comment by hramezani** [comment]
Thanks @Smixi

**Pull request merged**